Add uploading support for registered users

oparoz opened the pull request:
Fixes #25. It replaces #485 and #520 .
### Featuring
- Drag and drop component
- Folder uploading on Chrome
- File comparison dialogue
- [ + ] button 
- Improved empty page, now with error messages
### Todo
- [x] Defer quota check in web UI when overwriting shared file owncloud/core@ceaefc2
### Test plan

**Base**
- Uploading file(s) via drag and drop
- Uploading a folder via drag and drop
- Uploading file(s) via the [+] menu

**Additional tests**
- Uploading from an empty Gallery should work and the UI should adapt to the content of the album
- Uploading the same file twice should bring up the conflict dialogue
### Tech review
- The javascript used in `files` is loaded as-is. A special file has been created to replace FileList and Files methods called by those scripts
- `newfilemenu.js` has been imported and modified to make it work with Gallery. We don't need more than the upload entry
- `OC.Upload._isReceivedSharedFile` is replaced with a method which does the lookup in the cache instead of in the HTML
- Some methods were moved from Gallery to GalleryView as it made more sense to have them there
- GalleryImage has 2 new properties: size and sharedWithUser
- The landing page for empty albums has been modified to take into consideration the context in which that page is called and can print the error message, if communicated to the frontend

_Note: Ideally, we should switch to webdav once uploading is supported and re-usable components for the file dialogue move outside of `files`._

---

@karlitschek @jancborchardt @DeepDiver1975 @LukasReschke @PVince81 @MorrisJobke @jospoortvliet @setnes @demattin @arkascha @elpraga @sualko @bugsbane @mmattel @tahaalibra @viraj96 @imjalpreet @mbtamuli @rahulgoyal030

oparoz committed:
Add uploading support for registered users

oparoz committed:
Add upload button
Based on the files + button
Allows uploading multiple files at once

oparoz committed:
Improve the uploading experience for smaller screens

oparoz committed:
Update the empty page to reflect the fact that the user can now upload directly

tahaalibra commented:
**Working**
- [x] Uploading file(s) via drag and drop
- [x] Uploading a folder via drag and drop
- [x] Uploading file(s) via the [+] menu
- [x] Uploading from an empty Gallery
- [x] UI is adapting to the changes
- [x] Uploading Same files gives the warning

great  :+1: ...everything seems to be working fine

I will try some more test and take a look at the code..

imjalpreet commented:
@oparoz Everything seems to be working! :+1: 
Except a few points that I will be mentioning at the end.
I have tested the following:
- Drag and Drop Upload for both file(s) and folder.
- Upload file(s) via [+] button.
- Uploading from an empty Gallery is working (and the UI adapts to the images and also to window resize)
- Conflict Dialogue comes if same files are uploaded more than once.
- Error Messages

But I wanted to clarify one thing, if we try to upload a folder and even if there is one empty file, the entire folder doesn't get uploaded. I think it should not be like this. It think we should try to atleast upload the non empty files and folders. 

Secondly, the error message that is displayed in the above case should be changed. Currently, the error message is 'Error uploading file "Untitled Document": Unable to upload Untitled Document as it is a directory or has 0 bytes'. Since, directory upload is now possible, we should not print in the error message that the reason of not uploading is that it is a directory. I am attaching a screenshot of the following.

One more thing we can do is automatically hide the error message after a certain time instead of keeping it till someone clicks on 'x'.
![errormessage](https://cloud.githubusercontent.com/assets/6432146/13660558/2ea017f4-e6b1-11e5-9875-78c89f674247.png)

tahaalibra commented:
I am getting javascript error when viewing a public gallery..it is preventing gallery top open
''galleryview.js?v=95a8dd6…:276 Uncaught TypeError: Cannot set property '_isReceivedSharedFile' of undefined"
![selection_001](https://cloud.githubusercontent.com/assets/6804341/13663931/72d4a052-e6ca-11e5-80ef-1ee587129604.jpg)

@oparoz

oparoz committed:
Detect if images presented are shared with the current user

imjalpreet commented:
@oparoz Now public gallery is also working fine! :+1:

oparoz commented:
> I am getting javascript error when viewing a public gallery..it is preventing gallery top open
> ''galleryview.js?v=95a8dd6…:276 Uncaught TypeError: Cannot set property '_isReceivedSharedFile' of undefined"

@tahaalibra - Ah, yes. I've changed approach at the 11th hour and failed to verify that it still worked on the public side. That side doesn't support uploading yet. It's now fixed.

@imjalpreet - All the problems you've described should happen on the Files side as well and some reports have probably already been opened in `core`. Go see if that's the case and if not, add your suggestions. Some of these issues are not trivial to fix, but if you have solutions, `core` is also accepting PRs :)

imjalpreet commented:
@oparoz Yeah Sure. I will have a look in core.

tahaalibra commented:
@oparoz now working on public as well as logged in !!!!!

jospoortvliet commented:
Awesome work @oparoz !
